components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/dao/GroupDAOImpl.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.role.mgt.core.dao;
+
+import org.wso2.carbon.database.utils.jdbc.NamedPreparedStatement;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.role.mgt.core.IdentityRoleManagementClientException;
+import org.wso2.carbon.identity.role.mgt.core.IdentityRoleManagementException;
+import org.wso2.carbon.identity.role.mgt.core.IdentityRoleManagementServerException;
+import org.wso2.carbon.identity.role.mgt.core.RoleConstants;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.role.mgt.core.RoleConstants.Error.INVALID_REQUEST;
+import static org.wso2.carbon.identity.role.mgt.core.RoleConstants.Error.UNEXPECTED_SERVER_ERROR;
+import static org.wso2.carbon.identity.role.mgt.core.dao.SQLQueries.GET_GROUP_ID_BY_NAME_SQL;
+import static org.wso2.carbon.identity.role.mgt.core.dao.SQLQueries.GET_GROUP_NAME_BY_ID_SQL;
+
+/**
+ * RoleDAO Implementation.
+ */
+public class GroupDAOImpl implements GroupDAO {
+
+    private static volatile GroupDAOImpl instance;
+
+    private GroupDAOImpl() {
+
+    }
+
+    public static GroupDAOImpl getInstance() {
+
+        if (instance == null) {
+            synchronized (RoleDAOImpl.class) {
+                if (instance == null) {
+                    instance = new GroupDAOImpl();
+                }
+            }
+        }
+        return instance;
+    }
```

Comment:
Should we have a singleton for DAOs ?

@mefarazath @emswbandara , @JKAUSHALYA

Reply:
We can avoid the singleton pattern IMO.

Move the DAO implementation to a private package and expose DAOs through a Factory class. WDYT?
Having a factory allows use to change the DAO implementation without impacting the consuming code.

Reply:
Ok. I'll add a DAO Factory and change the implementation accordingly.